README.md
                  COMPRESSED LOOPBACK BLOCK DEVICE (cloop.ko)
=============================================================================

DISCLAIMER: THE AUTHORS HEREBY DENY ANY WARRANTY FOR ANY DATA LOSS AND
DAMAGE WHATSOEVER THAT MAY BE CAUSED DIRECTLY OR INDIRECTLY BY USING THIS
SOFTWARE. THIS IS EXPERIMENTAL SOFTWARE. USE AT YOUR OWN RISK.

This module is licensed under the GNU GENERAL PUBLIC LICENSE Version 2, as
stated in the source.

-----------------------------------------------------------------------------

This is cloop, a Kernel module to add support for filesystem-independent,
transparently decompressed, read-only block devices.

Original Author: Paul 'Rusty' Russel
Extensions, Bugfixes & newer Kernels: Klaus Knopper (http://knopper.net/knoppix/)

INSTALLATION:
-------------

make KERNEL_DIR=/path/to/linux-kernel/sources
(as root) mkdir -p /lib/modules/misc && cp cloop.o /lib/modules/misc/ 
(as root) depmod -a
(as root) mknod /dev/cloop b 240 0
(as root) mknod /dev/cloop1 b 240 1
...

USAGE:
------

Creating a compressed image:
 create_compressed_fs image blocksize > image.cloop_compressed

blocksize must be a multiple of 512  bytes. Make sure you have enough
swap to hold the entire compressed image in virtual memory! Use "-"
as filename to read data from stdin, as in:

 mkisofs -r datadir | create_compressed_fs - 65536 > datadir.iso.compressed

Mounting a compressed image (see above for device creation):
 insmod cloop.o file=/path/to/compressed/image
 mount -o ro -t whatever /dev/cloop /mnt/compressed

Starting from cloop version 1.0, setting of cloop compressed files via
losetup /dev/cloop1 /path/to/file
is possible. cloop 1.0 uses 64bit pointers instead of 32bit, so cloop files
are no longer limited to <= 4GB (uncompressed size).

cloop supports the losetup ioctls for adding and removing files via /dev/cloop*

For more information, please refer to the sources. If you don't understand
what all this is about, please DON'T EVEN ATTEMPT TO INSTALL OR USE THIS
SOFTWARE.

A download location for the currentmost version of this module and utilities
has been provided at: https://github.com/KlausKnopper/cloop

		-Klaus Knopper 26.08.2014 <cloop@knopper.net>


debian/changelog
cloop (3.14-1) unstable; urgency=medium

  * Updates for Kernel >= 3.14
  * Preparing for git repository upload

 -- Klaus Knopper <knoppix@knopper.net>  Tue, 26 Aug 2014 17:14:37 +0200

cloop (2.639-2) unstable; urgency=low

  * Change compressed_loop.h -> cloop.h everywhere.

 -- Klaus Knopper <knoppix@knopper.net>  Wed, 06 Jul 2011 18:43:20 +0200

cloop (2.639-1) unstable; urgency=low

  * Changes for kernel 2.6.39, replace lock_kernel() by mutex_lock()

 -- Klaus Knopper <knoppix@knopper.net>  Tue, 10 May 2011 14:18:21 +0200

cloop (2.637-1) unstable; urgency=low

  * Add compat_ioctl for compilation in 64bit mode.

 -- Klaus Knopper <knoppix@knopper.net>  Tue, 11 Jan 2011 22:57:31 +0100

cloop (2.636-1) unstable; urgency=low

  * Changes for new kernel 2.6.36 removed block layer api function and
    locked_ioctl.

 -- Klaus Knopper <knoppix@knopper.net>  Mon, 25 Oct 2010 22:40:25 +0200

cloop (2.634-1) unstable; urgency=low

  * Changes for new kernel 2.6.34 sector limit functions
  * Set device read-only by default

 -- Klaus Knopper <knoppix@knopper.net>  Wed, 25 Aug 2010 13:36:01 +0200

cloop (2.631-1) unstable; urgency=low

  * Changes for new kernel 2.6.31 block device handling

 -- Klaus Knopper <knoppix@knopper.net>  Tue, 15 Sep 2009 21:26:57 +0200

cloop (2.629-1) unstable; urgency=low

  * Acquired Eduard Blochs new advfs.cc and extract_compressed_fs.c

 -- Klaus Knopper <knoppix@knopper.net>  Wed, 15 Apr 2009 10:59:32 +0200

cloop (2.628-2) unstable; urgency=low

  * Reflect kernel 2.6.28 block device API changes in order to actually work.

 -- Klaus Knopper <knoppix@knopper.net>  Tue, 04 Nov 2008 18:38:02 +0100

cloop (2.628-1) unstable; urgency=low

  *  Minor source modifications for compiling with kernel 2.6.28

 -- Klaus Knopper <knoppix@knopper.net>  Sat, 25 Oct 2008 23:11:39 +0200

cloop (2.625-1) unstable; urgency=low

  * Kernel 2.6.25/26 api changes

 -- Klaus Knopper <knoppix@knopper.net>  Mon, 19 May 2008 18:44:27 +0200

cloop (2.624-1) unstable; urgency=low

  * Kernel 2.6.24 api change (rq_for_each_segment)

 -- Klaus Knopper <knoppix@knopper.net>  Sat, 26 Jan 2008 13:05:45 +0100

cloop (2.623-1) unstable; urgency=low

  * Fixed alloc_disk-leak (caused modprobe cloop; rmmod cloop; modprobe cloop to fail).
  * Added support for busybox losetup.

 -- Klaus Knopper <knoppix@knopper.net>  Sun, 16 Dec 2007 23:10:01 +0200

cloop (2.622-1) unstable; urgency=low

  * Rewrite with Kernel-Thread and request-queue for Kernel 2.6.22 and up.
  * Abandoned Kernel 2.4 support.
  * Replaced obsolete advancecomp-1.9 by 1.15

 -- Klaus Knopper <knoppix@knopper.net>  Sun, 23 Sep 2007 22:19:05 +0200

cloop (2.06-2) unstable; urgency=low

  * Patch for Kernel 2.6.22 invalidate_bdev(one arg).

 -- Klaus Knopper <knoppix@knopper.net>  Tue, 10 Jul 2007 06:16:26 +0200

cloop (2.06-1) unstable; urgency=low

  * Added "cloop_suspend" ioctl and utility contributed by Fabian Franz.

 -- Klaus Knopper <knoppix@knopper.net>  Sun, 25 Feb 2007 22:25:52 +0100

cloop (2.05-1) unstable; urgency=low

  * Adopted Eduard Blochs debian buildscripts for generic cloop 2.05

 -- Klaus Knopper <knoppix@knopper.net>  Sun, 20 Aug 2006 14:10:24 +0200

cloop (2.04-1+eb.1-7) unstable; urgency=low

  * disabled debugging output (ooops...)
  * another ugly bug identified by Andreas Krüger (thanks!), broken offset
    index generation in tempfile mode (closes: #382762)
  * Makefile fixes to adapt to new advfs.cc path
  * manpage patch from Andreas to explain extract_compressed_fs usage applied
    (closes: #382765) and SGML format fixes

 -- Eduard Bloch <blade@debian.org>  Sun, 13 Aug 2006 12:43:44 +0200

cloop (2.04-1+eb.1-6) unstable; urgency=low

  * updated advancecomp to 1.15, better documenting (diff) modifications used
    to implant advfs.cc
  * bugfixes in advfs.cc:
    - recognice -m/-f with standard output (closes: #382611)
    - compressing block selection, could get stuck in rare cases

 -- Eduard Bloch <blade@debian.org>  Sat, 12 Aug 2006 13:26:53 +0200

cloop (2.04-1+eb.1-5) unstable; urgency=low

  * not using the install-module rule from the Makefile, installing module
    explicitely (closes: #381639)
  * added debian/compat to the modules tarball
  * fixed broken macros in Millan's patch (closes: #381626)

 -- Eduard Bloch <blade@debian.org>  Sun, 06 Aug 2006 19:04:28 +0200

cloop (2.04-1+eb.1-4) unstable; urgency=low

  * Dirty little optimisation of the block-for-compression selection

 -- Eduard Bloch <blade@debian.org>  Wed,  2 Aug 2006 23:53:05 +0200

cloop (2.04-1+eb.1-3) unstable; urgency=low

  * rewrite of the work control in multithreaded advfs/create_compressed_fs,
    now definitely not adding a zero block on aligned data (closes: #296179)
  * fixed a memory leak in "best" mode and better control for memory
    allocation now (finally closes: #164015 and hopefully closes: #380384,
    reopen if it doesn't)
  * added pt.po (closes: #337546)
  * added GNU/kFreeBSD porting patch from Robert Millan (closes: #335872)

 -- Eduard Bloch <blade@debian.org>  Wed, 02 Aug 2006 23:51:14 +0200

cloop (2.04-1+eb.1-2) unstable; urgency=low

  * removed obsolete CFLAGS from the Makefile, confused current kernel build
    system, plus some user hints added.
  * compressed_loop.c: removed all conditional blocks with workarounds for
    kernel << 2.6.8, including kernel 2.4 related junk. Gets the correct
    module name now and no compiler warnings anymore. Fixed intendation.
  * debian/control: removed some old cruft, updated to debhelper 5, updated
    Standards-Version.

 -- Eduard Bloch <blade@debian.org>  Fri, 28 Jul 2006 19:17:18 +0200

cloop (2.04-1+eb.1-1) unstable; urgency=low

  * Sync my modified version with 2.04.1
    + fixes for 2.6.17 changes (closes: #376713)
  * fixes for compilatin with 2.6.18-rc2 (disabling devfs)
  * -s -> -S in create_compressed_fs help output (closes: #360764)

 -- Eduard Bloch <blade@debian.org>  Thu, 27 Jul 2006 19:19:30 +0200


cloop_suspend.c
/* 
 * cloop_suspend - Suspend a cloop device until losetup /dev/cloop <file> is 
 *                 run again.
 * 
 * Copyright (c) 2007 by Fabian Franz.
 *
 * License: GPL, v2.
 * 
 */

#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>

/* We don't use the structure, so that define does not hurt */
#define dev_t int
#include <linux/loop.h>
#include "cloop.h"

int main(int argc, char** argv)
{
	if (argc < 2)
	{
		fprintf(stderr, "syntax: %s <device>\n", argv[0]);
		fprintf(stderr, "        suspends and unlocks cloop <device>\n");
		fprintf(stderr, "        until a new file is loaded via losetup\n");
		return 1;
	}
	int fd = open(argv[1], O_RDONLY);

	if (fd < 0)
	{
		perror(argv[1]);
		return 1;
	}

	if (ioctl(fd, CLOOP_SUSPEND) < 0)
	{
		perror("ioctl: CLOOP_SUSPEND");
		return 1;
	}

	close(fd);

	return 0;
}


extract_compressed_fs.c
/* Extracts a filesystem back from a compressed cloop file */
/* Extended to support stdin 31.5.2008 Klaus Knopper       */
/* License: GPL V2                                         */

#define _LARGEFILE64_SOURCE
#define _XOPEN_SOURCE 600

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <endian.h>
#include <errno.h>
#include <string.h>
#include <zlib.h>
#include <netinet/in.h>
#include <inttypes.h>

#ifdef __CYGWIN__
typedef uint64_t loff_t;
#endif
#ifndef be64toh
static __inline __uint64_t
__bswap64(__uint64_t _x)
{

	return ((_x >> 56) | ((_x >> 40) & 0xff00) | ((_x >> 24) & 0xff0000) |
	    ((_x >> 8) & 0xff000000) | ((_x << 8) & ((__uint64_t)0xff << 32)) |
	    ((_x << 24) & ((__uint64_t)0xff << 40)) |
	    ((_x << 40) & ((__uint64_t)0xff << 48)) | ((_x << 56)));
}
#if BYTE_ORDER == LITTLE_ENDIAN
#define be64toh(x)	__bswap64(x)
#else
#define be64toh(x) x
#endif
#endif /* !be64toh */
#define __be64_to_cpu be64toh
#include "cloop.h"

struct compressed_block
{
	size_t size;
	void *data;
};

int main(int argc, char *argv[])
{
	int handle, output;
	unsigned int i, total_blocks, total_offsets, offsets_size,
	    compressed_buffer_size, uncompressed_buffer_size;
	struct cloop_head head;
	unsigned char *compressed_buffer, *uncompressed_buffer;
	loff_t *offsets;
	/* For statistics */
	loff_t compressed_bytes, uncompressed_bytes, block_modulo;

	if (argc != 3) {
		fprintf(stderr, "Syntax: %s infile outfile, use \"-\" for stdin/stdout.\n", argv[0]);
		exit(1);
	}

	if(!strcmp(argv[1],"-")) handle = STDIN_FILENO;
	else {
		handle = open(argv[1], O_RDONLY|O_LARGEFILE);
		if (handle < 0) {
			perror("Opening compressed input file\n");
			exit(1);
		}
		/* Never ever attempt to cache file content in
		 * filesystem cache, since we really need it just ONCE. */
		fdatasync(handle);
		posix_fadvise(handle, 0, 0, POSIX_FADV_DONTNEED|POSIX_FADV_SEQUENTIAL);
	}

	if(!strcmp(argv[2],"-")) output = STDOUT_FILENO;
	else {
		output = open(argv[2], O_CREAT|O_WRONLY|O_LARGEFILE,
		                       S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
		if (output < 0) {
			perror("Opening uncompressed output file\n");
			exit(1);
		}
		/* Never ever attempt to cache file content in
		 * filesystem cache, since we really need it just ONCE. */
		fdatasync(output);
		posix_fadvise(output, 0, 0, POSIX_FADV_DONTNEED|POSIX_FADV_SEQUENTIAL);
	}


	if (read(handle, &head, sizeof(head)) != sizeof(head)) {
		perror("Reading compressed file header\n");
		exit(1);
	}

	total_blocks = ntohl(head.num_blocks);
	uncompressed_buffer_size = ntohl(head.block_size);

	fprintf(stderr, "%s: compressed input has %u blocks of size %u.\n",
		argv[0], total_blocks, uncompressed_buffer_size);


	/* The maximum size of a compressed block, due to the
	 * specification of uncompress() */
	compressed_buffer_size = uncompressed_buffer_size + uncompressed_buffer_size/1000 + 12 + 4;
	compressed_buffer = malloc(compressed_buffer_size);
	if (compressed_buffer == NULL) {
		perror("Out of memory for compressed buffer");
		fprintf(stderr," (%d bytes).\n", compressed_buffer_size);
		exit(1);
	}

	uncompressed_buffer = malloc(uncompressed_buffer_size);
	if (uncompressed_buffer == NULL) {
		perror("Out of memory for uncompressed buffer");
		fprintf(stderr," (%d bytes).\n", uncompressed_buffer_size);
		exit(1);
	}


	/* Store block index in memory to avoid seek()ing a lot */
	total_offsets  = total_blocks + 1;
	offsets_size = total_offsets * sizeof(loff_t);
	offsets = (loff_t *)malloc(offsets_size);
	if (offsets == NULL) {
		perror("Out of memory");
		fprintf(stderr, " for %d offsets.\n", total_offsets);
		exit(1);
	}

	if (read(handle, offsets, offsets_size) != offsets_size) {
		perror("Reading offsets");
		fprintf(stderr, " (%d bytes).\n", offsets_size);
		exit(1);
	}
	
	for (i = 0, compressed_bytes=0, uncompressed_bytes=0, block_modulo = total_blocks / 10;
	     i < total_blocks;
	     i++) {
                int size = __be64_to_cpu(offsets[i+1]) - __be64_to_cpu(offsets[i]);
		uLongf destlen = uncompressed_buffer_size;
		if (size < 0 || size > compressed_buffer_size) {
			fprintf(stderr, 
				"%s: Size %d for block %u (offset %" PRIu64 ") wrong, corrupt data!\n",
				argv[0], size, i, (uint64_t) __be64_to_cpu(offsets[i]));
			exit(1);
		}
		if(read(handle, compressed_buffer, size) != size) {
			perror("Reading block");
			fprintf(stderr, " %u (offset %" PRIu64 ") of size %d.\n", i,
			     (uint64_t) __be64_to_cpu(offsets[i]), size);
			exit(1);
		}

#if 0 /* DEBUG */
		if (i == 3) {
			fprintf(stderr,
				"Block head:%02X%02X%02X%02X%02X%02X%02X%02X\n",
				buffer[0],
				buffer[1],
				buffer[2],
				buffer[3],
				buffer[4],
				buffer[5],
				buffer[6],
				buffer[7]);
			fprintf(stderr,
				"Block tail:%02X%02X%02X%02X%02X%02X%02X%02X\n",
				buffer[3063],
				buffer[3064],
				buffer[3065],
				buffer[3066],
				buffer[3067],
				buffer[3068],
				buffer[3069],
				buffer[3070]);
		}
#endif
		switch (uncompress(uncompressed_buffer, &destlen,
				   compressed_buffer, size)) {
			case Z_OK: break;

			case Z_MEM_ERROR:
				fprintf(stderr, "Uncomp: oom block %u\n", i);
				exit(1);
				break;

			case Z_BUF_ERROR:
				fprintf(stderr, "Uncomp: not enough out room %u\n", i);
				exit(1);
				break;

			case Z_DATA_ERROR:
				fprintf(stderr, "Uncomp: input corrupt %u\n", i);
				exit(1);
				break;

			default:
				fprintf(stderr, "Uncomp: unknown error %u\n", i);
				exit(1);
		}
		compressed_bytes += size; uncompressed_bytes += destlen;
		if(((i % block_modulo) == 0) || (i == (total_blocks - 1))) {
			fprintf(stderr, "[Current block: %6u, In: %" PRIu64 "kB, Out: %" PRIu64 "kB, ratio %d%%, complete %3d%%]\n",
			        i, 
              (uint64_t) compressed_bytes / 1024L,
              (uint64_t) uncompressed_bytes / 1024L,
				(int)((uncompressed_bytes * 100L) / compressed_bytes),
				(int)(i * 100 / (total_blocks - 1)));
		}
		write(output, uncompressed_buffer, destlen);
		fdatasync(output);
	}
	return 0;
}


create_compressed_fs.c
/* Creates a compressed image, given a file as an argument.
 * (c)1999 Paul `Rusty' Russell.  GPL.
 *
 * CHANGELOG:
 * * Sun Okt 26 01:05:29 CEST 2003 Klaus Knopper
 * - Changed format of index pointers to network byte order
 * * Sat Sep 29 2001 Klaus Knopper <knopper@knopper.net>
 * - changed compression to Z_BEST_COMPRESSION,
 * * Sat Jun 17 2000 Klaus Knopper <knopper@knopper.net>
 * - Support for reading file from stdin,
 * - Changed Preamble.
 * * Sat Jul 28 2001 Klaus Knopper <knopper@knopper.net>
 * - cleanup and gcc 2.96 / glibc checking
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#if defined(__FreeBSD__)
#include <sys/endian.h>
#include <netinet/in.h>
typedef uint64_t loff_t;
#ifndef htobe64
static __inline __uint64_t
__bswap64(__uint64_t _x)
{

	return ((_x >> 56) | ((_x >> 40) & 0xff00) | ((_x >> 24) & 0xff0000) |
	    ((_x >> 8) & 0xff000000) | ((_x << 8) & ((__uint64_t)0xff << 32)) |
	    ((_x << 24) & ((__uint64_t)0xff << 40)) |
	    ((_x << 40) & ((__uint64_t)0xff << 48)) | ((_x << 56)));
}
#if BYTE_ORDER == LITTLE_ENDIAN
#define htobe64(x)	__bswap64(x)
#else
#define htobe64(x)
#endif
#endif
#define __cpu_to_be64 htobe64
#else
#include <asm/byteorder.h>
#endif
#include <fcntl.h>
#include <zlib.h>
#include "compressed_loop.h"

#define MAX_KMALLOC_SIZE 2L<<17

#define CLOOP_PREAMBLE "#!/bin/sh\n" "#V2.0 Format\n" "insmod cloop.o file=$0 && mount -r -t iso9660 /dev/cloop $1\n" "exit $?\n"

struct cb_list
{
	struct cb_list *next;
	size_t size;
	char data[0];
};

void free_cb_list(struct cb_list *cbl)
{
 if(cbl->next) free_cb_list(cbl->next);
 free(cbl);
}

/* Now using the goto style because it is quicker to read */
static struct cb_list *create_compressed_blocks(int handle, unsigned long
                          blocksize, unsigned long *numblocks)
{
 struct cb_list *cbl,**cbp=&cbl;
 unsigned long i=0;
 unsigned int last;
 unsigned long long total_uncompressed=0,total_compressed=0;
 unsigned long maxlen = blocksize + blocksize/1000 + 12;
 char *compressed, *uncompressed;
 if((uncompressed=malloc(blocksize))==NULL)
  {
   fprintf(stderr, "*** Can't malloc(%ld).\n",blocksize);
   return NULL;
  }
 if((compressed=malloc(maxlen))==NULL)
  {
   fprintf(stderr, "*** Can't malloc(%ld).\n",blocksize);
   goto free_uncompressed;
  }
 for(i=0,last=0; !last; i++)
  {
   int z_error;
   unsigned long total=0, len = maxlen;
   memset(compressed,0,len); memset(uncompressed,0,blocksize);
   while(total<blocksize) /* Read a complete block */
    {
     ssize_t r=read(handle, uncompressed+total, blocksize-total);
     if(r<=0) { last=1; break; }
     total+=r;
    }
   total_uncompressed += total;
   if (total != blocksize)
    {
     last=1;
     fprintf(stderr, "Partial read (%lu bytes of %lu), padding with zeros.\n",
					total, blocksize);
    }
   if((z_error=compress2(compressed, &len, uncompressed, blocksize, Z_BEST_COMPRESSION)) != Z_OK)
    {
     fprintf(stderr, "*** Error %d compressing block %lu! (compressed=%p, len=%lu, uncompressed=%p, blocksize=%lu)\n", z_error, i, compressed,len,uncompressed,blocksize);
     goto error_free_cb_list;
    }
   if((*cbp = malloc(sizeof(struct cb_list)+len))==NULL) /* get another block */
    {
     fprintf(stderr, "*** Out of memory allocating block ptrs (virtual memory exhausted).\n");
     goto error_free_cb_list;
    }
   total_compressed+=len;
   /* Print status */
   fprintf(stderr, "Block# %5lu size %6lu -> %6lu [compression ratio %3lu%%, overall: %3Lu%%]\n", i, total, len, total>0?((len*100)/total):100,total_uncompressed>0?((total_compressed*100)/total_uncompressed):100);
   (*cbp)->size = len;
   memcpy((*cbp)->data, compressed, len);
   (*cbp)->next=NULL;
   cbp=&((*cbp)->next);
  } /* for */
 goto free_compressed;

 error_free_cb_list:
    if(cbl) { free_cb_list(cbl); cbl=NULL; i=0; }

 free_compressed:
    free(compressed);
 free_uncompressed:
    free(uncompressed);
 
 *numblocks=i;
 return cbl;
}

int main(int argc, char **argv)
{
 int in;
 unsigned long blocksize;
 struct cloop_head head;
 unsigned long numblocks;
 unsigned long long bytes_so_far;
 unsigned long i;
 struct cb_list *compressed_blocks,*cbp;

 if (argc != 3)
  {
   fprintf(stderr, "Usage: %s filename blocksize(bytes).\n",argv[0]);
   fprintf(stderr, "Use '-' as filename for stdin.\n");
   return 1;
  }

 blocksize = atoi(argv[2]);
 if (blocksize == 0 || blocksize % 512 != 0)
  {
   fprintf(stderr, "*** Blocksize must be a multiple of 512.\n");
   return 1;
  }

 if (blocksize > MAX_KMALLOC_SIZE)
  {
   fprintf(stderr, "WARNING: Blocksize %lu may be too big for a kmalloc() (%lu max).\n",blocksize,MAX_KMALLOC_SIZE);
   sleep(2);
  }

 if (sizeof(CLOOP_PREAMBLE) > CLOOP_HEADROOM)
  {
   fprintf(stderr, "*** Preamble (%u chars) > headroom (%u)\n",
			sizeof(CLOOP_PREAMBLE), CLOOP_HEADROOM);
   return 1;
  }
		
 in=strcmp(argv[1],"-")==0?dup(fileno(stdin)):open(argv[1], O_RDONLY);

 if (in < 0)
  {
   perror("Opening input");
   return 1;
  }

 compressed_blocks = create_compressed_blocks(in, blocksize, &numblocks);

 close(in);

 memset(head.preamble, 0, sizeof(head.preamble));
 memcpy(head.preamble, CLOOP_PREAMBLE, sizeof(CLOOP_PREAMBLE));
 head.block_size = htonl(blocksize);
 head.num_blocks = htonl(numblocks);

 fprintf(stderr, "Block size %lu, number of blocks %lu.\n",
         blocksize, numblocks);

 bytes_so_far = sizeof(head) + sizeof(loff_t) * (numblocks + 1);

 /* Write out head... */
 write(STDOUT_FILENO, &head, sizeof(head));

 if (!compressed_blocks) return 1;

 /* Write offsets */
 for (i=0,cbp=compressed_blocks; i < numblocks+1; i++)
  {
   loff_t tmp;
   tmp = __cpu_to_be64(bytes_so_far);
   write(STDOUT_FILENO, &tmp, sizeof(tmp));
   if(cbp) { bytes_so_far += cbp->size; cbp=cbp->next; }
  }

 /* Now write blocks and free them. */
 for (i = 0, cbp=compressed_blocks; cbp && i < numblocks; i++)
  {
   if (write(STDOUT_FILENO, cbp->data, cbp->size) != cbp->size)
    {
     perror("writing block");
     free_cb_list(compressed_blocks);
     return 1;
    }
   cbp=cbp->next;
   free(compressed_blocks); compressed_blocks=cbp;
  }
#if defined(__FreeBSD__)
 /*
  * FreeBSD requires padding to 512 byte boundary
  */
 bytes_so_far = lseek(STDOUT_FILENO, 0, SEEK_END);
 if (bytes_so_far % 512)
  {
   static char padding[512];
   off_t len = 512 - bytes_so_far % 512;

   if (write(STDOUT_FILENO, padding, len) != len)
    {
     perror("writing padding block");
     return 1;
    }
  }
#endif
 fprintf(stderr,"Done.\n");
 return 0;
}
